Profile patching - Invalid memory address or nil pointer dereference

Running `skaffold deploy -p <profile>` throws a SIGSEGV error without giving a clue on what has caused it.

Configuration errors should be expected and handled gracefully.

### Expected behavior
An informative error output, ideally pointing to the line that has caused the error, if any.

### Actual behavior
skaffold deploy -p dev

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x11ae6cc]

goroutine 1 [running]:
github.com/GoogleContainerTools/skaffold/vendor/github.com/krishicks/yaml-patch.tryReplace(0x2527c80, 0xc0000b8068, 0xc0002c0180, 0x20bd020, 0xc00024c390)
	vendor/github.com/krishicks/yaml-patch/operation.go:110 +0x18c
github.com/GoogleContainerTools/skaffold/vendor/github.com/krishicks/yaml-patch.(*Operation).Perform(0xc0002c0180, 0x2527c80, 0xc0000b8068, 0x1, 0xffffffffffffffff)
	vendor/github.com/krishicks/yaml-patch/operation.go:72 +0x272
github.com/GoogleContainerTools/skaffold/vendor/github.com/krishicks/yaml-patch.Patch.Apply(0xc00056e0e0, 0x3, 0x4, 0xc000028000, 0x18d, 0x200, 0x2, 0x4, 0x11, 0xc00020d4f8, ...)
	vendor/github.com/krishicks/yaml-patch/patch.go:52 +0x5b1
github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema.applyProfile(0xc0001e4000, 0xc0003208f0, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	pkg/skaffold/schema/profiles.go:167 +0x66b
github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema.ApplyProfiles(0xc0001e4000, 0x32ee2a0, 0x1672401, 0x2503580)
	pkg/skaffold/schema/profiles.go:50 +0x226
github.com/GoogleContainerTools/skaffold/cmd/skaffold/app/cmd.newRunner(0x32ee2a0, 0xc0000e3c10, 0x2527740, 0xc0002c1f00, 0xc0000e3c10)
	cmd/skaffold/app/cmd/runner.go:46 +0x9f
github.com/GoogleContainerTools/skaffold/cmd/skaffold/app/cmd.run(0x24f0720, 0xc0000b8000, 0x0, 0x0)
	cmd/skaffold/app/cmd/run.go:55 +0xa1
github.com/GoogleContainerTools/skaffold/cmd/skaffold/app/cmd.NewCmdDeploy.func1(0xc00036c000, 0xc0000cfba0, 0x0, 0x2, 0x0, 0x0)
	cmd/skaffold/app/cmd/deploy.go:34 +0x60
github.com/GoogleContainerTools/skaffold/vendor/github.com/spf13/cobra.(*Command).execute(0xc00036c000, 0xc0000cfb80, 0x2, 0x2, 0xc00036c000, 0xc0000cfb80)
	vendor/github.com/spf13/cobra/command.go:762 +0x465
github.com/GoogleContainerTools/skaffold/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x32d8560, 0xc0000b8000, 0x24f0720, 0xc0000bc000)
	vendor/github.com/spf13/cobra/command.go:852 +0x2ec
github.com/GoogleContainerTools/skaffold/vendor/github.com/spf13/cobra.(*Command).Execute(...)
	vendor/github.com/spf13/cobra/command.go:800
github.com/GoogleContainerTools/skaffold/cmd/skaffold/app.Run(0x0, 0x0)
	cmd/skaffold/app/skaffold.go:32 +0x8a
main.main()
	cmd/skaffold/skaffold.go:29 +0x26
```

### Information

- Skaffold version: 0.25
- MacOS
- Contents of skaffold.yaml:

```yaml

apiVersion: skaffold/v1beta7
kind: Config
build:
  artifacts:
    - image: gcr.io/<project>/backend
      jibGradle: {}
      context: app
    - image: gcr.io/<project>/frontend
      docker:
        dockerfile: docker/images/node
        buildArgs:
          API_BASE_URL: "http://<base url>/api"
      context: webapp
deploy:
  kustomize:
    path: kustomize/overlays/dev

profiles:
  - name: dev
    patches:
      - op: replace
        path: /build/artifacts/0/image/
        value: gcr.io/<project id>/dev-backend
      - op: replace
        path: /build/artifacts/1/image/
        value: gcr.io/<project id>/dev-frontend
      - op: replace
        path: /build/artifacts/1/docker/buildArgs/0/API_BASE_URL/
        value: "http://<base url>/api"
```

--- a/pkg/skaffold/schema/profiles.go
+++ b/pkg/skaffold/schema/profiles.go
@@ -156,12 +156,18 @@ func applyProfile(config *latest.SkaffoldPipeline, profile latest.Profile) error
 			op = "replace"
 		}
 
-		patches = append(patches, yamlpatch.Operation{
+		patch := yamlpatch.Operation{
 			Op:    yamlpatch.Op(op),
 			Path:  yamlpatch.OpPath(patch.Path),
 			From:  yamlpatch.OpPath(patch.From),
 			Value: patch.Value,
-		})
+		}
+
+		if !tryPatch(patch, buf) {
+			return fmt.Errorf("invalid path: %s", patch.Path)
+		}
+
+		patches = append(patches, patch)
 	}
 
 	buf, err = yamlpatch.Patch(patches).Apply(buf)
@@ -170,6 +176,20 @@ func applyProfile(config *latest.SkaffoldPipeline, profile latest.Profile) error
 	}
 
 	return yaml.Unmarshal(buf, config)
+}
+
+// tryPatch is here to verify patches one by one before we
+// apply them because yamlpatch.Patch is known to panic when a path
+// is not valid.
+func tryPatch(patch yamlpatch.Operation, buf []byte) (valid bool) {
+	defer func() {
+		if errPanic := recover(); errPanic != nil {
+			valid = false
+		}
+	}()
+
+	_, err := yamlpatch.Patch([]yamlpatch.Operation{patch}).Apply(buf)
+	return err == nil
 }
 
 func profilesByName(profiles []latest.Profile) map[string]latest.Profile {

--- a/pkg/skaffold/schema/profiles_test.go
+++ b/pkg/skaffold/schema/profiles_test.go
@@ -70,6 +70,34 @@ profiles:
 	testutil.CheckDeepEqual(t, "Dockerfile.second", pipeline.Build.Artifacts[1].DockerArtifact.DockerfilePath)
 }
 
+func TestApplyInvalidPatch(t *testing.T) {
+	config := `build:
+  artifacts:
+  - image: example
+profiles:
+- name: patches
+  patches:
+  - path: /build/artifacts/0/image/
+    value: replacement
+`
+
+	tmp, cleanup := testutil.NewTempDir(t)
+	defer cleanup()
+
+	yaml := fmt.Sprintf("apiVersion: %s\nkind: Config\n%s", latest.Version, config)
+	tmp.Write("skaffold.yaml", yaml)
+
+	parsed, err := ParseConfig(tmp.Path("skaffold.yaml"), false)
+	testutil.CheckError(t, false, err)
+
+	pipeline := parsed.(*latest.SkaffoldPipeline)
+	err = ApplyProfiles(pipeline, &cfg.SkaffoldOptions{
+		Profiles: []string{"patches"},
+	})
+
+	testutil.CheckErrorAndDeepEqual(t, true, err, "appying profile patches: invalid path: /build/artifacts/0/image/", err.Error())
+}
+
 func TestApplyProfiles(t *testing.T) {
 	tests := []struct {
 		description string

